src/Tablebot/Plugins/Quote.hs

```diff
@@ -270,29 +309,158 @@ deleteQ qId m =
 renderCustomQuoteMessage t (Quote txt author submitter msgId cnlId dtm) qId m = do
-  guild <- findGuild m
+  guild <- contextGuildId m
   let link = getLink guild
-  void $
-    sendEmbedMessage
-      m
-      t
-      ( addColour Blue $
-          addTimestamp dtm $
-            addFooter (pack $ "Quote #" ++ show qId) $
-              simpleEmbed (txt <> "\n - " <> author <> maybeAddFooter link)
-      )
+  return
+    ( (messageDetailsBasic t)
+        { messageDetailsEmbeds =
+            Just
+              [ addColour DiscordColorBlue $
+                  addTimestamp dtm $
+                    addFooter (pack $ "Quote #" ++ show qId) $
+                      simpleEmbed (txt <> "\n - " <> author <> maybeAddFooter link)
+              ]
+        }
+    )
   where
     getLink :: Maybe GuildId -> Maybe Text
     getLink = fmap (\x -> getMessageLink x (fromIntegral cnlId) (fromIntegral msgId))
     maybeAddFooter :: Maybe Text -> Text
     maybeAddFooter (Just l) = "\n[source](" <> l <> ") - added by " <> submitter
     maybeAddFooter Nothing = ""
 
+quoteApplicationCommand :: CreateApplicationCommand
+quoteApplicationCommand = CreateApplicationCommandChatInput "quote" "store and retrieve quotes" (Just opts) True
+  where
+    opts =
+      ApplicationCommandOptionsSubcommands $
+        ApplicationCommandOptionSubcommandOrGroupSubcommand
+          <$> [ addQuoteAppComm,
+                showQuoteAppComm,
+                randomQuoteAppComm,
+                authorQuoteAppComm,
+                editQuoteAppComm
+              ]
+    addQuoteAppComm =
+      ApplicationCommandOptionSubcommand
+        "add"
+        "add a new quote"
+        [ ApplicationCommandOptionValueString "quote" "what the actual quote is" True (Left False),
+          ApplicationCommandOptionValueString "author" "who authored this quote" True (Left False)
+        ]
+    showQuoteAppComm =
+      ApplicationCommandOptionSubcommand
+        "show"
+        "show a quote by number"
+        [ ApplicationCommandOptionValueInteger "id" "the quote's number" True (Left True) (Just 1) Nothing
+        ]
+    randomQuoteAppComm =
+      ApplicationCommandOptionSubcommand
+        "random"
+        "show a random quote"
+        []
+    authorQuoteAppComm =
+      ApplicationCommandOptionSubcommand
+        "author"
+        "show a random quote by an author"
+        [ApplicationCommandOptionValueString "author" "whose quotes do you want to see" True (Left False)]
+    editQuoteAppComm =
+      ApplicationCommandOptionSubcommand
+        "edit"
+        "edit a quote"
+        [ ApplicationCommandOptionValueInteger "quoteid" "the id of the quote to edit" True (Left False) Nothing Nothing,
+          ApplicationCommandOptionValueString "quote" "what the actual quote is" False (Left False),
+          ApplicationCommandOptionValueString "author" "who authored this quote" False (Left False)
+        ]
+
+quoteApplicationCommandRecv :: Interaction -> DatabaseDiscord ()
```

Comment:
This definition makes it look like you have to put in a hell of a lot of work to get interactions to work. Is there a way this could be reused/reduced? As I say in the conclusion of this review, I'd like to see a tutorial about interactions and slash commands - maybe when you write this tutorial it'll become clear how things could be simplified.

Reply:
so the issue is that our current subcommanding turns things into the normal text-parsing straight away, which means I've had to manually implement subcommands here.  Additionally, due to how interactions work, I've had to make the quote adding and quote editing "self referential", meaning that the "view quote" link is the bot's message itself, which is obviously a bit tricky.

I also have some rudimentary slash command auto complete implemented for `/quote show`, which isn't very clean either.